### PR TITLE
copr: support collation aware compare_in

### LINKED ISSUE #6575: Implement collation aware built-in functions for Coprocessor
## Feature Request

This is a follow up for https://github.com/tikv/tikv/issues/6574. Coprocessor built-in functions need to be updated to respect collations.

- [x] LT/LE/GT/GE/EQ/NE/NULLEQString
- [x] Like
- [ ] InString

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

###  What have you changed?

support collation aware `compare_in`

 Resolve #6575

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?

- Unit test

### CODE REVIEW

**Comment by andylokandy** [comment]
I implemented some utils in https://github.com/tikv/tikv/pull/6714, which might be helpful for your PR.

**Comment by andylokandy** [comment]
I implemented some utils in https://github.com/tikv/tikv/pull/6714, which might be useful for your PR.

**Comment by andylokandy** [comment]
If you use things in https://github.com/tikv/tikv/pull/6714, you can remove `InByHash` and transmute the `HashSet` as a whole like this.https://github.com/tikv/tikv/blob/fa24ebdce6807ed4e337e4026aa473fafbece2aa/components/tidb_query/src/batch/executors/fast_hash_aggr_executor.rs#L280-L282

**Comment by TennyZhuang** [comment]
> If you use things in #6714, you can remove `InByHash` and transmute the `HashSet` as a whole like this.
> 
> https://github.com/tikv/tikv/blob/fa24ebdce6807ed4e337e4026aa473fafbece2aa/components/tidb_query/src/batch/executors/fast_hash_aggr_executor.rs#L280-L282

`InByHash` is used to share the same code between Bytes and other types. It seems that I can replace `CollationAwareBytes` with `SortKey`.

**Comment by andylokandy** [comment]
I still think it's possible. Just check this out:
https://github.com/tikv/tikv/blob/fa24ebdce6807ed4e337e4026aa473fafbece2aa/components/tidb_query/src/batch/executors/fast_hash_aggr_executor.rs#L360-L395

File: components/tidb_query/src/rpn_expr/impl_compare_in.rs [hunk 1]
```diff
@@ -2,16 +2,60 @@
...
 use tipb::{Expr, ExprType};
 
+use crate::codec::collation::*;
 use crate::codec::data_type::*;
 use crate::codec::mysql::{Decimal, MAX_FSP};
 use crate::{Error, Result};
 
+pub trait InByHash {
+    type Key: Evaluable + Extract + Eq;
+    type StoreKey: 'static + Hash + Eq + Sized + std::marker::Send;
```
Comment (Fullstop000) [inline components/tidb_query/src/rpn_expr/impl_compare_in.rs]: May add `use std::marker::Send` at L5 and just use `Send` here

File: components/tidb_query/src/rpn_expr/impl_compare_in.rs [hunk 2]
```diff
@@ -2,16 +2,60 @@
...
+
+    fn map(key: Self::Key) -> Result<Self::StoreKey>;
+    fn map_ref(key: &Self::Key) -> Result<&Self::StoreKey>;
+}
+
+pub struct NormalInByHash<K: Evaluable + Extract + Hash + Eq + Sized + std::marker::Send>(
```
Comment (Fullstop000) [inline components/tidb_query/src/rpn_expr/impl_compare_in.rs]: ditto

File: components/tidb_query/src/rpn_expr/impl_compare_in.rs [hunk 3]
```diff
@@ -2,16 +2,60 @@
...
+    fn map(key: Self::Key) -> Result<Self::StoreKey> {
+        SortKey::new(key).map_err(Error::from)
+    }
+
+    fn map_ref(key: &Self::Key) -> Result<&Self::StoreKey> {
+        SortKey::new_ref(key).map_err(Error::from)
+    }
+}
+
 pub trait Extract: std::marker::Sized {
```
Comment (Fullstop000) [inline components/tidb_query/src/rpn_expr/impl_compare_in.rs]: ditto

**Comment by TennyZhuang** [comment]
@sre-bot wake up

**Comment by you06** [comment]
@Fullstop000 @TennyZhuang I'll fix the approve issue later today.

**Comment by you06** [comment]
@Fullstop000 you should be able to use `/approve` and `/approve cancel` now.

**Pull request merged**